### DISCUSSION TITLE
Why is `functools.partial` used so frequently?

### DISCUSSION THREAD

**Discussion by jordandsullivan** [unitaryfund/mitiq#2326] [Q&A] [answered]
In a fair number of example notebooks and tests, the utility `functools.partial` is used to creator an executor function with fewer arguments specific arguments set to particular values. e.g. https://github.com/unitaryfund/mitiq/blob/main/docs/source/examples/ddd_tutorial.md?plain=1#L182. 

If this is such a common use case (creating executors which accept a smaller number of arguments), why not advise users to use optional positional arguments with sensible defaults? Why the need to create a whole new executor function using `functools.partial`?

**Comment by jordandsullivan** [comment, marked as answer]
Upon further reflection, I suppose it is a convenient way to specify positional arguments that will stay the same across many executions of a function.

**Reply by natestemen** [reply]
Yup that's exactly it!

Since users can pass both python functions, and `Executor` objects into Mitiq, we could potentially add a method that returns a new executor with the specified `kwargs`. E.g. this could look like the following.

```py
def execute(circuit, noise_level=100, shots=1000):
    ...

executor = Executor(execute)

executor_with_more_shots = executor.set_params(shots=10_000)
```

I think _most_ users of Mitiq are now using an `Executor` object, however, as it's not required.

**Reply by jordandsullivan** [reply]
I may be wrong, but it doesn't look like the usage of `partial` is explained anywhere in our documentation, although it is used in multiple examples. I think it might be useful to add a short description and example to the [Executors](https://mitiq.readthedocs.io/en/stable/guide/executors.html) page in our docs.  

Created an issue for it: https://github.com/unitaryfund/mitiq/issues/2360

**Reply by natestemen** [reply]
Makes sense. Since we have an issue open to track this now, feel free to close this discussion.
